W-22549992: Document recommendations on prompt stream consumption (#3354)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/streaming-about.adoc

----

[[consume-streams-promptly]]
== Consume Streams Promptly

When a connector operation returns a stream (such as reading a file, querying a database, or receiving an HTTP response body), the stream holds underlying resources open (file handles, database cursors, network connections) until it is fully consumed or the event that originated the stream completes processing.

=== Why Prompt Consumption Matters

* *Resource exhaustion*: Unconsumed streams keep resources (connections, file descriptors, cursors) allocated. If many streams accumulate without being consumed, the application can run out of available resources, eventually leading to `OutOfMemoryError` or connection pool exhaustion.
* *Memory pressure*: With repeatable streaming (the default), the runtime buffers stream content in memory or on disk. Large unconsumed streams occupy buffer space until the event completes.
* *Connection starvation*: For HTTP and database operations, the underlying connection cannot be returned to the pool until the response stream is consumed or the event finishes.

=== Recommendations

[%header,cols="1,2"]
|===
| Practice | Explanation

| Always consume or transform the payload

> | Use a DataWeave transformation, `<logger>` with payload access, or write to a target to ensure the stream is read to completion.

Reviewer: I've seen apps trying to consume the payload with a `<set-payload value="#[payload] />`, which is more similar to a noop. Should we clarify it or assume the dev knows that?

 </sfdc:query>
----

[[consume-streams-promptly]]
== Consume Streams Promptly

> When a connector operation returns a stream (for example, when reading a file, querying a database, or receiving an HTTP response body), the stream keeps underlying resources open (file handles, database cursors, and network connections) until it is fully consumed or until processing of the originating event completes.

Reviewer 2: ```suggestion
When a connector operation returns a stream (for example, when reading a file, querying a database, or receiving an HTTP response body), the stream keeps underlying resources open (file handles, database cursors, and network connections) until it's fully consumed or until processing of the originating event completes.
```

=== Why Prompt Consumption Matters

* Resource exhaustion: Unconsumed streams keep resources (connections, file descriptors, and cursors) allocated. If many streams accumulate without being consumed, the application can run out of available resources and eventually hit `OutOfMemoryError` or connection pool exhaustion.
* Memory pressure: With repeatable streaming (the default), the runtime buffers stream content in memory or on disk. Large unconsumed streams occupy buffer space until the event completes.

> * Connection starvation: For HTTP and database operations, the underlying connection cannot return to the pool until the response stream is consumed or the event finishes.

Reviewer 2: ```suggestion
* Connection starvation: For HTTP and database operations, the underlying connection can't return to the pool until the response stream is consumed or the event finishes.
```

=== Recommendations

This section describes best practices for consuming streams and minimizing resource retention.

[%header,cols="1,2"]
|===
| Practice | Explanation

| Always consume or transform the payload
| Use a DataWeave transformation, a `<logger>`component that accesses the payload, or a target variable to ensure the stream is fully consumed.

| Avoid storing streams in variables without consuming them
| If you assign a stream to a variable but never read it, the underlying resources remain open until the event completes.

| Use `target` and `targetValue` appropriately

> | When using `target` to store operation results, ensure the stored value is eventually consumed, or use `targetValue` to force consumption at store time (for example, with a DataWeave expression that reads the stream).

Reviewer 2: ```suggestion
| When using `target` to store operation results, make sure the stored value is eventually consumed, or use `targetValue` to force consumption at store time (for example, with a DataWeave expression that reads the stream).
```
